Mathlib/Algebra/MonoidAlgebra/Degree.lean

```diff
@@ -232,15 +232,86 @@ using the maximum of the exponents. -/
     (f + g).supDegree D ≤ (f.supDegree D) ⊔ (g.supDegree D) :=
   sup_support_add_le D f g
 
-variable [CovariantClass B B (· + ·) (· ≤ ·)] [CovariantClass B B (Function.swap (· + ·)) (· ≤ ·)]
-  (D : A →+ B) in
-theorem supDegree_mul_le (f g : R[A]) :
-    (f * g).supDegree D ≤ f.supDegree D + g.supDegree D :=
-  sup_support_mul_le (fun {_ _} => (AddMonoidHom.map_add D _ _).le) f g
+@[simp]
+theorem supDegree_neg {f : R'[A]} :
+    (-f).supDegree D = f.supDegree D := by
+  rw [supDegree, supDegree, Finsupp.support_neg]
+
+theorem supDegree_sub_le {f g : R'[A]} :
+    (f - g).supDegree D ≤ f.supDegree D ⊔ g.supDegree D := by
+  rw [sub_eq_add_neg, ← supDegree_neg (f := g)]; apply supDegree_add_le
+
+theorem supDegree_sum_le {ι} {s : Finset ι} {f : ι → R[A]} :
+    (∑ i in s, f i).supDegree D ≤ s.sup (fun i => (f i).supDegree D) :=
+  (Finset.sup_mono Finsupp.support_finset_sum).trans_eq (Finset.sup_biUnion _ _)
+
+theorem supDegree_single_ne_zero (a : A) {r : R} (hr : r ≠ 0) :
+    (single a r).supDegree D = D a := by
+  rw [supDegree, Finsupp.support_single_ne_zero a hr, Finset.sup_singleton]
+
+theorem supDegree_single (a : A) (r : R) :
+    (single a r).supDegree D = if r = 0 then ⊥ else D a := by
+  split_ifs with hr <;> simp [supDegree_single_ne_zero, hr]
+
+variable {p q : R[A]}
+
+@[simp]
+theorem supDegree_zero : (0 : R[A]).supDegree D = ⊥ := by simp [supDegree]
+
+theorem ne_zero_of_supDegree_ne_bot : p.supDegree D ≠ ⊥ → p ≠ 0 := mt (fun h => h ▸ supDegree_zero)
+
+theorem ne_zero_of_not_supDegree_le {b : B} (h : ¬ p.supDegree D ≤ b) : p ≠ 0 :=
+  ne_zero_of_supDegree_ne_bot (fun he => h <| he ▸ bot_le)
+
+theorem apply_eq_zero_of_not_le_supDegree {a : A} (hlt : ¬ D a ≤ p.supDegree D) : p a = 0 := by
+  contrapose! hlt
+  exact Finset.le_sup (Finsupp.mem_support_iff.2 hlt)
+
+variable (hadd : ∀ a1 a2, D (a1 + a2) = D a1 + D a2)
+
+theorem supDegree_mul_le [CovariantClass B B (· + ·) (· ≤ ·)]
+    [CovariantClass B B (Function.swap (· + ·)) (· ≤ ·)] :
+    (p * q).supDegree D ≤ p.supDegree D + q.supDegree D :=
+  sup_support_mul_le (fun {_ _} => (hadd _ _).le) p q
+
+theorem supDegree_prod_le {R A B : Type*} [CommSemiring R] [AddCommMonoid A] [AddCommMonoid B]
+    [SemilatticeSup B] [OrderBot B]
+    [CovariantClass B B (· + ·) (· ≤ ·)] [CovariantClass B B (Function.swap (· + ·)) (· ≤ ·)]
+    {D : A → B} (hzero : D 0 = 0) (hadd : ∀ a1 a2, D (a1 + a2) = D a1 + D a2)
+    {ι} {s : Finset ι} {f : ι → R[A]} :
+    (∏ i in s, f i).supDegree D ≤ ∑ i in s, (f i).supDegree D := by
+  refine' s.induction _ _
+  · rw [Finset.prod_empty, Finset.sum_empty, one_def, supDegree_single]
+    split_ifs; exacts [bot_le, hzero.le]
+  · intro i s his ih
+    rw [Finset.prod_insert his, Finset.sum_insert his]
+    exact (supDegree_mul_le hadd).trans (add_le_add_left ih _)
+
+variable [CovariantClass B B (· + ·) (· < ·)] [CovariantClass B B (Function.swap (· + ·)) (· < ·)]
+
+theorem apply_add_of_supDegree_eq (hD : D.Injective) {ap aq : A}
```

Comment:
I've definitely seen some form of this lemma.  This probably means that you can use this lemma to golf a similar proof somewhere in the `Polynomial.coeff_mul_of_degree_le`-range (made up name, but I have to go now!).